商品出品機能

choragon0511 opened the pull request:
# what
商品出品機能作成

# why
商品を出品するため

ログイン状態で出品画面に行けるがログアウト状態だとログイン画面に行く
https://gyazo.com/d5633275a1aec86ff9e082e088f940cc

入力できてないとエラーが出る
https://gyazo.com/028afa7da6e8f20ffd7426bb4ec5d67e

入力できてると保存されトップページに行く
https://gyazo.com/9c0a167efa66541c0da85cea3736f5e7

販売利益が計算される
https://gyazo.com/8da98e6b3bc2788485e088290c74b666

itemモデルのテスト成功
https://gyazo.com/2512dc69c3416126e9aae4cd1abc6f36

choragon0511 committed:
itemモデル作成

itemモデル作成
マイグレーションファイル修正
rails db:migrate

choragon0511 committed:
出品画面への遷移

フリマ素材入れて
リンク編集
ログインしてる時に移動
ログインしてなかったらログイン画面に移動

choragon0511 committed:
Active Storageの導入

Active Storageの導入

choragon0511 committed:
ActiveHash導入

ActiveHash導入
カテゴリ選択できるように設定

choragon0511 committed:
出品保存

必要な情報を登録したら保存される
抜けがあるとページにとどまる
エラーメッセージが表示される

choragon0511 committed:
バリデーション

バリデーション設定

choragon0511 committed:
販売利益

JavaScriptで販売利益が計算されて
表示されるようにする

choragon0511 committed:
itemモデルテスト

Rspecでitemモデルのテスト書いた

TakashiKozuma commented on app/controllers/application_controller.rb:
```diff
@@ -1,4 +1,5 @@
 class ApplicationController < ActionController::Base
+  before_action :authenticate_user!, except: [:index]
```
```rb
before_action :authenticate_user!, except: [:index]
```
現状だと、今後実装する全てのコントローラーに対して適用されてしまいます。
items_controllerに記述を移して、適用される機能を限定しましょう！

TakashiKozuma commented on app/controllers/items_controller.rb:
```diff
@@ -1,6 +1,29 @@
 class ItemsController < ApplicationController
 
   def index
+    @items = Item.all
```
```rb
@items = Item.all
```
こちらの記述は商品一覧表示機能のものなのでいったんコメントアウトしておきましょう。
Githubのプルリクエストは差分のみが表示される仕様になっています。
そのためこのままだと一覧表示機能を実装した時にレビュワーにこのコードを見せられなくなってしまうためです。

TakashiKozuma commented on app/models/item.rb:
```diff
@@ -0,0 +1,20 @@
+  belongs_to :burden
+  belongs_to :prefecture
+  belongs_to :waitday
+
```
READMEにあるように、userとのアソシエーションを追加しましょう！

TakashiKozuma commented on app/views/items/new.html.erb:
```diff
@@ -0,0 +1,161 @@
+
+    <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+    <%= render 'shared/error_messages', model: f.object %>
+    <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
```rb
<%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
12,14行目のコメントアウトは実装案内なので削除しておきましょう。
実装案内は完成したアプリを人に見せる際に残しておくの望ましくないためです。
他にも「〜しましょう」のような実装案内があれば、実装時に削除するようにしていきましょう。

TakashiKozuma commented on spec/factories/items.rb:
```diff
@@ -0,0 +1,16 @@
+    prefecture_id {'2'}
+    wait_day_id {'2'}
+    price {'300'}
+
```
userとのアソシエーションを設定したら、userと紐づけるような記述を追記しましょう。
また、その後userと紐づいていないと保存できないこともテストしましょう。

TakashiKozuma commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,78 @@
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Prefecture can't be blank")
+      end
+      it "wait_day_idが空では登録できない" do
+        @item.wait_day_id = ""
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Wait day can't be blank")
+      end
```
```rb
      it "wait_day_idが空では登録できない" do
        @item.wait_day_id = ""
        @item.valid?
        expect(@item.errors.full_messages).to include("Wait day can't be blank")
      end
```
アクティブハッシュは何も選択していない場合は、空ではなく、1が選択された状態です。
そのため、空の文字列ではなく「1」が入力されていると保存できないことを確認するテストに変更しましょう。
他のアクティブハッシュのカラムについても同様です。

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
コメントを致しましたので、ご確認お願いいたします。

修正後はコミット・プッシュのし忘れがないかご確認の上、再度フォームからレビュー依頼をお願い致します。

choragon0511 committed:
修正

コントローラー
userとitemモデルのアソシエーション
コメント削除
FactoryBotのitemにuserと紐づけ
テストで紐付けがないと登録できない
アクティブハッシュは１で登録できない

choragon0511 commented:
コントローラー
userとitemモデルのアソシエーション
コメント削除
FactoryBotのitemにuserと紐づけ
テストで紐付けがないと登録できない
アクティブハッシュは１で登録できない
を修正

div-reviewer commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,83 @@
+        expect(@item.errors.full_messages).to include("Description can't be blank")
+      end
+      it "category_idが空では登録できない" do
+        @item.category_id = "1"
```
【任意修正】

```ruby
@item.category_id = "1"
↓
@item.category_id = 1
```

金額やActivehash等のinteger型に対しては文字列ではなく数値を代入する方がより綺麗な実装となるため、""は外してテストしてあげるようにしましょう。

div-reviewer commented on app/models/burden.rb:
```diff
@@ -0,0 +1,11 @@
+  self.data = [
+    { id: 1, name: '---' },
+    { id: 2, name: '着払い（購入者負担）' },
+    { id: 3, name: '送料込み（出品者負担）' },
```
【任意修正】

```ruby
 { id: 3, name: '送料込み（出品者負担）' },
```

,はデータが後ろに続く場合につけてあげましょう。
シンタックスエラー等、不要なエラーを起こさないようにするためです。
Rubocopを実行していただくと解消できるかと思われます。
既に実行はして頂いているかと思われますが、手動で修正が大変場合はこちらを再度実行しましょう。

div-reviewer reviewed (commented):
choragon0511様、学習お疲れ様です！
お忙しい中商品出品の修正頂きありがとうございました。

前回コメントについて綺麗に修正できているかと思われます。
よってこちら【LGTM】です。お疲れ様でした。

●任意修正について●
こちらリファクタリング（コード整理）や+aの修正の為、任意で修正して頂ければと思います。
修正された場合はこちら`再度レビュー依頼は出さなくて大丈夫です`ので、修正後、コミットとプッシュして頂けましたら、
マージ後、ローカルのmasterブランチに忘れずにプルし、次の実装に取り組んで頂ければと思います。

今後の実装も応援しております！どうぞよろしくおねがいします。

choragon0511 committed:
robocop

robocop
